M2/Macaulay2/d/sets.dd

```diff
@@ -19,6 +19,7 @@ makeSet(e:Expr):Expr := (
      is v:Sequence do makeSet(v)
      is w:List do if ancestor(w.Class,visibleListClass) then makeSet(w.v)
      else WrongArg("a visible list")
+     is h:HashTable do makeSet(keys(h))
```

Comment:
Could you add a typical value comment so `methods set` also lists `(set, HashTable)`?

Reply:
Oh, actually, this is worse than `set Set := identity` because it allocates new memory, no?

Reply:
yeah I'll rewrite it more cleanly, more like top-level `applyValues(h,x->1)`

Reply:
Actually, an easier solution might be to make it so people can install methods on set!

Reply:
yeah I don't know about that. we've been doing that a lot (superseding the d function, e.g., `toList` has suffered this fate) but I don't think it's great, it's not very clean. I'd rather basic functions like `set` be at d level.

Reply:
What is not "clean" about this? There are lots of constructions in mathematics that produce a set from an object. It seems very reasonable to allow package writers to define a type and say how `set T` should produce a set.

Reply:
what's not clean is having a d function, then renaming it at the m2 level, then introducing a new m2 function which shares the same name as the original d function but which potentially behaves differently. it's not good practice and can make debugging confusing.

Reply:
Another common approach is to give the d function a slightly different name, often appending "0" to the name.  So `set` could be the method function, and some of the installed methods could be the compiled function `set0`.

Reply:
Suppose you could declare methods in the interpreter and directly install new method functions on them in the top level. Would this be confusing? I agree that renaming the function and assigning a method is kind of hacky (and the `set0` trick is not much better), but the solution to that is to make it possible to declare methods in the interpreter, not prevent `set` from becoming a method in the top level!

Reply:
> Suppose you could declare methods in the interpreter and directly install new method functions on them in the top level.
yeah that'd be nice... but unless I'm mistaken this isn't currently possible. I was trying to keep this PR to a minimum.

Reply:
Yes, it's not possible. My point was that what you called unclean is a kludge until we have this feature, not that you should implement it now :)

Anyway, I think I'd like set to become a method, either in this PR or another time.